src/Hint/NewType.hs

```diff
@@ -65,3 +74,25 @@ hasNoStrategy _                      = False
 withNewtype :: Deriving a -> Deriving a
 withNewtype (Deriving l Nothing rs)  = Deriving l (Just $ DerivNewtype l) rs
 withNewtype d                        = d
+
+-- | Given a declaration, returns the suggested \"newtype\"ized declaration following these guidelines:
+-- * @MagicHash@'d stuff is __ignored__ - @data X = X Int#@
+-- * @ExistentialQuantification@ stuff is __ignored__ - @data X = forall t. X t@
+-- * Single field constructors get newtyped - @data X = X Int@ -> @newtype X = X Int@
+-- * Single record field constructors get newtyped - @data X = X {getX :: Int}@ -> @newtype X = X {getX :: Int}@
+-- * All other declarations are ignored.
+--
+-- TODO: insert ! warning (or lack thereof) somewhere
+singleSimpleFieldNew :: Hs.HsDecl pass -> Maybe (Hs.HsDecl pass)
+singleSimpleFieldNew (Hs.TyClD wtfIsThisSomePassStuff decl@(Hs.DataDecl _ name _ _ def@(Hs.HsDataDefn _ Hs.DataType ctx _ _ [Hs.L _ constructor] derives)))
```

Comment:
I can only guess what [`XTyClD p`](https://hackage.haskell.org/package/ghc-8.6.5/docs/HsExtension.html#t:XTyClD) from the `TyClD` constructor (and all other similar `X*` stuff) is.

Reply:
All the `X` things are for trees-that-grow extension points. Always just match with `_` and ignore them.

Reply:
Generally use `ext` for all `X` things, and `noExt` if you need to create one. Being polymorphic in the `pass` is usually not a good idea - the error messages get much worse and you don't gain any real flexibility - you probably `GhcPs` as `pass`.